plugin/src/main/java/org/opensearch/ml/processor/MLInferenceIngestProcessor.java

```diff
@@ -0,0 +1,411 @@
+        String modelId,
+        List<Map<String, String>> inputMaps,
+        List<Map<String, String>> outputMaps,
+        Map<String, String> modelConfigMaps,
+        int maxPredictionTask,
+        String tag,
+        String description,
+        boolean ignoreMissing,
+        ScriptService scriptService,
+        Client client
+    ) {
+        super(tag, description);
+        this.inferenceProcessorAttributes = new InferenceProcessorAttributes(
+            modelId,
+            inputMaps,
+            outputMaps,
+            modelConfigMaps,
+            maxPredictionTask
+        );
+        this.ignoreMissing = ignoreMissing;
+        this.scriptService = scriptService;
+        this.client = client;
+    }
+
+    /**
+     * overwrite in this execute method,
+     * when batch inference is available,
+     * to support async multiple predictions.
+     */
+    @Override
+    public void execute(IngestDocument ingestDocument, BiConsumer<IngestDocument, Exception> handler) {
+
+        List<Map<String, String>> processInputMap = inferenceProcessorAttributes.getInputMaps();
+        List<Map<String, String>> processOutputMap = inferenceProcessorAttributes.getOutputMaps();
+        int num_of_prediction_tasks = (processInputMap != null) ? processInputMap.size() : 0;
+
+        GroupedActionListener<Void> batchPredictionListener = new GroupedActionListener<>(new ActionListener<Collection<Void>>() {
+            @Override
+            public void onResponse(Collection<Void> voids) {
+                handler.accept(ingestDocument, null);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                handler.accept(null, e);
+            }
+        }, Math.max(num_of_prediction_tasks, 1));
+
+        for (int i = 0; i < Math.max(num_of_prediction_tasks, 1); i++) {
+            processPredictions(ingestDocument, batchPredictionListener, processInputMap, processOutputMap, 0, num_of_prediction_tasks);
+        }
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument ingestDocument) throws Exception {
+        throw new UnsupportedOperationException("this method should not get executed.");
+    }
+
+    /**
+     * process predictions for one model for multiple rounds of predictions
+     * ingest documents after prediction rounds are completed,
+     * when no input mappings provided, default to add all fields to model input fields,
+     * when no output mapping provided, default to output as
+     * "inference_results" field (the same format as predict API)
+     * @param ingestDocument The IngestDocument object containing the data to be processed.
+     * @param batchPredictionListener The GroupedActionListener for batch prediction.
+     * @param processInputMap A list of maps containing input field mappings.
+     * @param processOutputMap A list of maps containing output field mappings.
+     * @param i The current index of the prediction task.
```

Comment:
Make the variable name more readable, `i` is not a good name

Reply:
changed to 
     * @param inputMapIndex The current index of the inputMap.
     * @param inputMapSize The size of inputMap.
     * 
https://github.com/opensearch-project/ml-commons/blob/b9dfc86ad0d656530d1f112d11704828691bd335/plugin/src/main/java/org/opensearch/ml/processor/MLInferenceIngestProcessor.java#L140